### PR TITLE
Check padding density by input of embedding module

### PR DESCRIPTION
**Pull request by guyang3532**
### Description
The PaddingElimination optimization is enabled when the density of embedding padding less than 90%. We need to check the density of the embedding padding to decide whether enable the optimization.

Before this pr, we just check the inputs of graph and correlate one with the embedding node by iterate graph from the embedding node back to one graph input.
This is hard to be general because there may be complicated pattern between graph input and embedding node.

This pr check padding density by the direct input of embedding module rather than the input of graph at the first graph execution when exporting onnx graph.
And if the density < 90%, insert a flag PythonOp after the embedding node as:
```
             Embedding
		  |
            PythonOp (func_name:_FlagPaddingElimination)   (insert if density < 90%)
		  |
            Following graph
```

When the PaddingElimination is invoked, it check if there is the flag PythonOp(func_name:_FlagPaddingElimination) after the Embedding node and if it is, remove it and do the padding elimination optimization.

### CODE REVIEW

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc [hunk 1]
```diff
@@ -17,8 +17,12 @@ namespace onnxruntime {
 namespace {
 
 // TODO(pengwa): remove this once customized PythonOp shape inference is supported.
-constexpr const char* kInspectActivationFuncName = "onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation";
-constexpr const char* kIncrementStepFuncName = "onnxruntime.training.utils.hooks._subscriber_manager._IncrementStep";
+constexpr const char* kInspectActivationFuncName =
+    "onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation";
+constexpr const char* kIncrementStepFuncName =
+    "onnxruntime.training.utils.hooks._subscriber_manager._IncrementStep";
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc:23]:
```suggestion
    "onnxruntime.training.utils.hooks._IncrementStep";
```

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 2]
```diff
@@ -622,6 +647,43 @@ def __setstate__(self, state):
 
         _utils.reinitialize_graph_execution_manager(self)
 
+    def _check_embedding_sparsity(self):
+        if not self._runtime_options.enable_embedding_sparse_optimizer or self._device.type != "cuda":
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:651]:
```suggestion
        if not self._runtime_options.enable_sparse_optimizer or not self._runtime_options.enable_embedding_sparse_optimizer or self._device.type != "cuda":
```

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 3]
```diff
@@ -622,6 +647,43 @@ def __setstate__(self, state):
...
+            return
+
+        def embedding_hook(module, args, output):
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:654]:
```suggestion
        def _embedding_hook(module, args, output):
```

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 4]
```diff
@@ -49,6 +49,30 @@ def __init__(self, state, output_info: List[Tuple[torch.Size, torch.device, torc
         self.output_info = output_info
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:51]: add comment for why we need this clas and how it works.?

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py [hunk 5]
```diff
@@ -253,6 +253,10 @@ def forward(self, *inputs, **kwargs):
             ):
                 self.time_tracker.start(ORTModuleInitPhase.EndToEnd)
 
+                # need to set device before check embedding sparsity because it uses device
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_training_manager.py:256]: can we move the check into _export_model in orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 6]
```diff
@@ -622,6 +647,43 @@ def __setstate__(self, state):
 
         _utils.reinitialize_graph_execution_manager(self)
 
+    def _check_embedding_sparsity(self):
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:650]: move this into runter_inspector?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 7]
```diff
@@ -49,6 +49,30 @@ def __init__(self, state, output_info: List[Tuple[torch.Size, torch.device, torc
         self.output_info = output_info
 
 
+class _FlagPaddingElimination(torch.autograd.Function):
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:52]: move to runtime_inspector.py?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 8]
```diff
@@ -622,6 +647,43 @@ def __setstate__(self, state):
...
+                return _FlagPaddingElimination.apply(output)
+            else:
+                self._logger.info("Embedding sparsity-based optimization is OFF for density: %.0f%%", embed_density)
+                return None
+
+        for name, sub_module in self._flattened_module.named_modules():
+            if isinstance(sub_module, torch.nn.modules.sparse.Embedding):
+                if sub_module.padding_idx is not None and sub_module.padding_idx >= 0:
+                    self._embedding_module_to_padding_density_map[sub_module] = [name, -1]
+                    sub_module.register_forward_hook(embedding_hook)
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:685]: better to remove the hook after the model export or exception catching branch if it is failed.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 9]
```diff
@@ -91,6 +115,7 @@ def __init__(
         # Inspector for runtime information, for example input data, memory usage, etc.
         self._runtime_inspector = RuntimeInspector(self._logger, self._original_module)
         self._runtime_inspector.memory_ob.enable_memory_stats_by_step(self._runtime_options.print_memory_stat_by_step)
+        self._embedding_module_to_padding_density_map = {}
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:118]: i would suggest to put this in runtime_inspector still.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 6 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:662]: This is not convenient to move into runtime_inspector because the runtime_inspector.input_density_ob is not initialized before export_model. Need more refactor.  I will move it into runtime_inspector in next pr which will do the loss label inspect automatically.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 8 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:697]: this hook must be added before model export.

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc [hunk 10]
```diff
@@ -17,8 +17,12 @@ namespace onnxruntime {
 namespace {
 
 // TODO(pengwa): remove this once customized PythonOp shape inference is supported.
-constexpr const char* kInspectActivationFuncName = "onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation";
-constexpr const char* kIncrementStepFuncName = "onnxruntime.training.utils.hooks._subscriber_manager._IncrementStep";
+constexpr const char* kInspectActivationFuncName =
+    "onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation";
```
[the remark this answers is not in the document]
Comment (guyang3532) [reply orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc]: fixed, thx,

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc (hunk 1 above)
Comment (guyang3532) [reply orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc]: fixed, thx

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 2 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: fixed, thx

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 3 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: fixed, thx

File: orttraining/orttraining/python/training/ortmodule/_training_manager.py (hunk 5 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_training_manager.py]: have moved into

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 7 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: moved, thx.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 9 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: yes, moved, thanks

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 4 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py]: done, thx

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc [hunk 11]
```diff
@@ -17,8 +17,12 @@ namespace onnxruntime {
 namespace {
 
 // TODO(pengwa): remove this once customized PythonOp shape inference is supported.
-constexpr const char* kInspectActivationFuncName = "onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation";
-constexpr const char* kIncrementStepFuncName = "onnxruntime.training.utils.hooks._subscriber_manager._IncrementStep";
+constexpr const char* kInspectActivationFuncName =
+    "onnxruntime.training.utils.hooks._statistics_subscriber._InspectActivation";
+constexpr const char* kIncrementStepFuncName =
+    "onnxruntime.training.utils.hooks._statistics_subscriber._IncrementStep";
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc:23]: this is not correct.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 12]
```diff
@@ -633,6 +638,51 @@ def __setstate__(self, state):
 
         _utils.reinitialize_graph_execution_manager(self)
 
+    def _add_check_embedding_sparsity_hook(self):
+        if (
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:642]: add doc string for what this function did?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 13]
```diff
@@ -633,6 +638,51 @@ def __setstate__(self, state):
...
+            valid_token = torch.count_nonzero(ebd_input - module.padding_idx)
+            total_token = ebd_input.numel()
+            embed_density = float(valid_token) / float(total_token) * 100
+            if embed_density < 90:
+                self._logger.info("Embedding sparsity-based optimization is ON for density: %.0f%%", embed_density)
+                if module not in self._runtime_inspector._embedding_module_to_padding_density_map:
+                    self._logger.warning("Found Embedding module not in the map. %s", module)
+                    return None
+                if (
+                    module in self._runtime_inspector._embedding_module_to_padding_density_map
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:664]:
```suggestion
                    
```

File: orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py [hunk 14]
```diff
@@ -747,3 +748,30 @@ def _get_user_config_without_freq(configs: str):
             return notes, mem_tbl
 
         return [], None
+
+
+# When sparsity of embedding padding is more than 10%, this class will added as a PythonOp after the Embedding Node
+# in the ONNX graph.
+# It will be used as a flag to tell the graph transformer to modify the graph to eliminate the embedding padding.
+class FlagPaddingElimination(torch.autograd.Function):
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_runtime_inspector.py:756]: Out docstring format looks like this
 """
    This class is used to run the subscriber's forward and backward functions.
    The function will be called by two kinds of callers:
        1. SubscriberManager calls it for each registered nn.Module.
        2. Users who want to inspect the activation tensor at any place of model definition code.
    """

File: orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc [hunk 15]
```diff
@@ -18,9 +18,9 @@ namespace {
 
 // TODO(pengwa): remove this once customized PythonOp shape inference is supported.
 constexpr const char* kInspectActivationFuncName =
-    "onnxruntime.training.utils.hooks._statistics_subscriber._InspectActivation";
+    "onnxruntime.training.utils.hooks._subscriber_manager._InspectActivation";
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc:21]:
```suggestion
    "onnxruntime.training.utils.hooks._statistics_subscriber._InspectActivation";
```

**Review by pengwa — APPROVED** [review]

**Review by pengwa — APPROVED** [review]
